File: app/src/ultimate/java/com/clockmods/widget/config/WidgetPreviewView.java
package com.clockmods.widget.config;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.RemoteViews;

import com.clockmods.widget.model.WidgetConfig;
import com.clockmods.widget.render.WidgetRemoteViewsFactory;
import com.clockmods.widget.render.WidgetSizeClassResolver;
import com.clockmods.widget.update.WidgetUpdateCoordinator;

/**
 * Renders the draft through the real {@code WidgetRemoteViewsFactory}, so what the user previews
 * is byte-for-byte the layout the Launcher will inflate. Drafts never enter the preference store.
 *
 * <p>The size class is resolved from this view's own box rather than from the screen, which keeps
 * the preview in the same size class — and therefore the same visible module set — as the
 * recommended cell footprint for the widget kind.
 */
public final class WidgetPreviewView extends FrameLayout {
    private volatile int generation;
    private WidgetConfig latest;
    private final Runnable renderLatest = this::render;

    public WidgetPreviewView(Context context) {
        super(context);
    }

    public void show(WidgetConfig config) {
        latest = config;
        generation++;
        removeCallbacks(renderLatest);
        // A slider may emit dozens of drafts per second. Only enqueue its latest settled frame;
        // widget updates and the Done transaction share the executor and must not queue behind it.
        postDelayed(renderLatest, 48);
    }

    private void render() {
        final WidgetConfig config = latest;
        if (config == null) return;
        final int current = generation;
        final float density = getResources().getDisplayMetrics().density;
        final float width = measuredPx(true) / density;
        final float height = measuredPx(false) / density;
        // RemoteViews refuses to invoke framework methods on AppCompat subclasses
        // ("view: androidx.appcompat.widget.AppCompatImageView can't use method with RemoteViews").
        // An Activity inflater always carries AppCompat's view factory, so the preview must be
        // inflated from the application context exactly like AppWidgetHostView does.
        final Context inflationContext = getContext().getApplicationContext().createConfigurationContext(
                new android.content.res.Configuration(getResources().getConfiguration()));
        WidgetUpdateCoordinator.execute(() -> {
            if (current != generation) return;
            RemoteViews views = WidgetRemoteViewsFactory.create(
                    inflationContext, config, WidgetSizeClassResolver.resolve(width, height));
            post(() -> {
                if (current != generation) return;
                removeAllViews();
                addView(views.apply(inflationContext, this),
                        new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));
                disableClicks(this);
            });
        });
    }

    /** Falls back to the requested dp box so the very first preview already uses the right size. */
    private int measuredPx(boolean horizontal) {
        int size = horizontal ? getWidth() : getHeight();
        if (size > 0) return size;
        ViewGroup.LayoutParams params = getLayoutParams();
        if (params != null) {
            size = horizontal ? params.width : params.height;
            if (size > 0) return size;
        }
        return Math.round((horizontal ? 340 : 156) * getResources().getDisplayMetrics().density);
    }

    private static void disableClicks(View view) {
        view.setOnClickListener(null);
        view.setClickable(false);
        if (view instanceof ViewGroup) {
            ViewGroup group = (ViewGroup) view;
            for (int i = 0; i < group.getChildCount(); i++) disableClicks(group.getChildAt(i));
        }
    }

    @Override protected void onDetachedFromWindow() {
        generation++;
        removeCallbacks(renderLatest);
        super.onDetachedFromWindow();
    }

    @Override protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight) {
        super.onSizeChanged(width, height, oldWidth, oldHeight);
        if (latest != null && (width != oldWidth || height != oldHeight)) show(latest);
    }
}

File: app/src/ultimate/java/com/clockmods/widget/config/WidgetPreviewView.kt
package com.clockmods.widget.config

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import com.clockmods.R
import com.clockmods.widget.model.WidgetConfig
import com.clockmods.widget.model.WidgetKind
import com.clockmods.widget.render.WidgetRemoteViewsFactory
import com.clockmods.widget.render.WidgetSizeClassResolver
import com.clockmods.widget.update.WidgetUpdateCoordinator
import kotlin.math.roundToInt

/** Compose host for the real RemoteViews preview used by the launcher. */
@Composable
internal fun WidgetPreview(
    config: WidgetConfig,
    modifier: Modifier = Modifier,
) {
    val desiredWidth: Dp
    val desiredHeight: Dp
    when (config.kind) {
        WidgetKind.DIGITAL -> {
            desiredWidth = 340.dp
            desiredHeight = 96.dp
        }
        WidgetKind.WEATHER -> {
            desiredWidth = 340.dp
            desiredHeight = 156.dp
        }
        else -> {
            desiredWidth = 176.dp
            desiredHeight = 176.dp
        }
    }

    BoxWithConstraints(
        modifier = modifier.fillMaxWidth(),
        contentAlignment = Alignment.Center,
    ) {
        val previewWidth = minOf(maxWidth, desiredWidth)
        val previewHeight = desiredHeight
        AndroidView(
            factory = { context ->
                WidgetPreviewView(context).apply { id = R.id.widget_config_preview }
            },
            update = { preview ->
                preview.show(config, previewWidth.value, previewHeight.value)
            },
            modifier = Modifier.width(previewWidth).height(previewHeight),
        )
    }
}

/** Bridges RemoteViews into Compose; no configuration controls live in this View. */
class WidgetPreviewView(context: Context) : FrameLayout(context) {
    private var generation = 0
    private var lastConfig: WidgetConfig? = null
    private var lastWidthDp = Float.NaN
    private var lastHeightDp = Float.NaN

    fun show(config: WidgetConfig) {
        val density = resources.displayMetrics.density
        show(
            config,
            measuredPx(horizontal = true) / density,
            measuredPx(horizontal = false) / density,
        )
    }

    internal fun show(config: WidgetConfig, widthDp: Float, heightDp: Float) {
        if (lastConfig === config && lastWidthDp == widthDp && lastHeightDp == heightDp) return
        lastConfig = config
        lastWidthDp = widthDp
        lastHeightDp = heightDp
        val current = ++generation
        val inflationContext = context.applicationContext
        WidgetUpdateCoordinator.execute {
            val views = WidgetRemoteViewsFactory.create(
                context,
                config,
                WidgetSizeClassResolver.resolve(widthDp, heightDp),
            )
            post {
                if (current != generation) return@post
                removeAllViews()
                addView(
                    views.apply(inflationContext, this),
                    LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT),
                )
                disableClicks(this)
            }
        }
    }

    private fun measuredPx(horizontal: Boolean): Int {
        var size = if (horizontal) width else height
        if (size > 0) return size
        layoutParams?.let { params ->
            size = if (horizontal) params.width else params.height
            if (size > 0) return size
        }
        return ((if (horizontal) 340 else 156) * resources.displayMetrics.density).roundToInt()
    }

    private fun disableClicks(view: View) {
        view.setOnClickListener(null)
        view.isClickable = false
        if (view is ViewGroup) {
            repeat(view.childCount) { disableClicks(view.getChildAt(it)) }
        }
    }

    override fun onDetachedFromWindow() {
        generation++
        lastConfig = null
        super.onDetachedFromWindow()
    }
}
